server/sonar-db-dao/src/main/java/org/sonar/db/component/BranchDao.java

```diff
@@ -49,6 +49,15 @@ public void upsert(DbSession dbSession, BranchDto dto) {
     }
   }
 
+  public void updateMainBranchName(DbSession dbSession, String projectUuid, String newBranchKey) {
```

Comment:
I think that any generic update method should not change the key. Also, it doesn't make much sense to have to give an entire branchDto to change the key. There should be a updateMainBranchName to change the row's key by querying the projectUuid. This makes the change atomic and makes sure that no other field that would be included in the BranchDto gets changed at the same time.